substrate/frame/support/procedural/src/stored.rs

```diff
@@ -0,0 +1,207 @@
+				field-bounding strategy.",
+			))
+		}
+	}
+}
+
+/// Main implementation of the `#[derive_stored]` macro.
+pub fn stored(
+	attr: proc_macro::TokenStream,
+	item: proc_macro::TokenStream,
+) -> proc_macro::TokenStream {
+	match stored_impl(attr.into(), item.into()) {
+		Ok(tokens) => tokens.into(),
+		Err(e) => e.to_compile_error().into(),
+	}
+}
+
+fn stored_impl(attr: TokenStream2, item: TokenStream2) -> Result<TokenStream2> {
+	let _args: StoredArgs = syn::parse2(attr)?;
+	let input: syn::DeriveInput = syn::parse2(item)?;
+
+	let field_types = match &input.data {
+		syn::Data::Struct(data_struct) => match &data_struct.fields {
+			syn::Fields::Named(fields) => fields.named.iter().map(|f| &f.ty).collect::<Vec<_>>(),
+			syn::Fields::Unnamed(fields) =>
+				fields.unnamed.iter().map(|f| &f.ty).collect::<Vec<_>>(),
+			syn::Fields::Unit => Vec::new(),
+		},
+		syn::Data::Enum(_) =>
+			return Err(Error::new(
+				input.span(),
+				"#[derive_stored] is only supported on structs, not enums",
+			)),
+		syn::Data::Union(_) =>
+			return Err(Error::new(
+				input.span(),
+				"#[derive_stored] is only supported on structs, not unions",
+			)),
+	};
+
+	// Collect all type parameters for scale_info skip_type_params.
+	// By default, we skip all type parameters in TypeInfo metadata as they're rarely needed.
+	let all_type_params: Vec<_> = input
+		.generics
+		.params
+		.iter()
+		.filter_map(|p| match p {
+			syn::GenericParam::Type(tp) => Some(&tp.ident),
+			_ => None,
+		})
+		.collect();
+
+	// Generate scale_info attribute to skip all type parameters
+	let scale_info_attr = if !all_type_params.is_empty() {
+		quote! {
+			#[scale_info(skip_type_params(#(#all_type_params),*))]
+		}
+	} else {
+		quote! {}
+	};
+
+	// Generate derive_where with field-based bounds
+	// This ensures consistent bounding strategy: bounds are applied to field types, not type
+	// parameters. Codec derives use their default strategy which also bounds fields automatically.
+	let derive_where_attr = if !field_types.is_empty() {
+		quote! {
+			#[derive_where(Clone, Eq, PartialEq, Debug; #(#field_types),*)]
+		}
+	} else {
+		// For unit structs, no field types to bound
+		quote! {
+			#[derive_where(Clone, Eq, PartialEq, Debug)]
+		}
+	};
+
+	let name = &input.ident;
+	let vis = &input.vis;
+	let generics = &input.generics;
+	let attrs = &input.attrs;
+
+	let body = match &input.data {
+		syn::Data::Struct(data_struct) => match &data_struct.fields {
+			syn::Fields::Named(fields) => {
+				let named = &fields.named;
+				quote! { { #named } }
+			},
+			syn::Fields::Unnamed(fields) => {
+				let unnamed = &fields.unnamed;
+				quote! { ( #unnamed ); }
+			},
+			syn::Fields::Unit => quote! { ; },
+		},
+		_ => unreachable!(
+			"input.data is already matched above for Struct/Enum/Union;\
+			all variants covered;\
+			qed"
+		),
+	};
+
+	Ok(quote! {
+		#derive_where_attr
+		#[derive(
+			::scale_info::TypeInfo,
```

Comment:
This requires that the crates are imported by the crate where this macro is called